### DISCUSSION TITLE
Dividing dxf into grids and plot them on png?

### DISCUSSION THREAD

**Discussion by melihpeker-pointr** [mozman/ezdxf#451] [Q&A] [answered]
Hi, thanks for this great library!

I want to divide a dxf into grids and save those grids one-by-one to a png file. To do this, I created different layouts for each grid. I want to print those layouts to png using the following code segment but instead it plots the whole dxf. I checked the layouts using AutoDesk Viewer and they seem fine. However, I couldn't get to print layouts on png. 

Is there a better way to save grids on different png files? If not so how can I plot different layouts to png files? Can you please help?

```python
layout2 = doc.layout('Layout2')
param = {"show_hatch": 0, "hatch_pattern": 0}
fig = plt.figure()
ax = fig.add_axes([0, 0, 1, 1])
ctx = RenderContext(doc, export_mode=False)
ctx.set_current_layout(layout2)
ctx.current_layout.set_colors(bg="#FFFFFFFF")
out = MatplotlibBackend(ax, params=param)
Frontend(ctx, out).draw_layout(layout2, finalize=True)
fig.savefig("Layout2.png", dpi=600)
```

The drawing is as this: 

<img width="682" alt="Screenshot 2021-06-01 at 12 52 59" src="https://user-images.githubusercontent.com/83574547/120305191-276bcc80-c2d9-11eb-8fac-d6ac8c856fa4.png">

And I wanted to save grids to png file like this (like zoomed version of patches):

<img width="686" alt="Screenshot 2021-06-01 at 12 53 04" src="https://user-images.githubusercontent.com/83574547/120305280-3d798d00-c2d9-11eb-82b4-771cf6e5cda1.png">

Instead I get this:

![Layout1](https://user-images.githubusercontent.com/83574547/120305350-4f5b3000-c2d9-11eb-8569-7cf3f6509bfa.png)

**Comment by mbway** [comment, marked as answer]
can you explain how you are constructing these additional zoomed in layouts? they seem to not be using viewports or you would not see the shapes in the image you posted (the drawing addon does not support viewports yet).
Are you setting the view boundary for the layout or something? the drawing addon completely ignores these because they are usually useless.
If you are able to calculate / find the view limits and you just want to set them on the axes after drawing then you can do:
```python
fig = plt.figure()
ax = fig.add_axes([0, 0, 1, 1])
...
Frontend(ctx, out).draw_layout(layout2, finalize=True)
ax.set_xlim(xmin, xmax)
ax.set_ylim(ymin, ymax)
```
you have to be careful to maintain the aspect ratio though. you can [adjust the figure size](https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure.set_size_inches) to maintain the correct aspect ratio.

**Reply by melihpeker-pointr** [reply]
I construct those paperspace layouts with the `page_setup` command:

```python
layout.page_setup(size=(15, 15), margins=(10, 15, 10, 15), units='mm', offset=(0, 0), rotation=0, scale=16,
                  name='ezdxf', device='DWG to PDF.pc3')
```

Also, I tried to use `zoom` function like 
```python
zoom.window(layout, (100, 100), (200, 200))
```
 but also it didn't help.

Btw, I can use the `set_xlim` commands of _matplotlib_ but I prefer to do that using layouts so that I can keep the margins and offsets information for further calculations on those grids. Thanks!

**Reply by melihpeker-pointr** [reply]
I guess `draw_layout` function does not care about zoom or page_setup. It only checks whether an entity is in that layout.

**Reply by mbway** [reply]
can't you do both though? as-in read out the view bounds of the layout and use it to set the axes limits? That's what would have to happen internally anyway if that was supported. Maybe @mozman has a better idea

**Reply by mbway** [reply]
there is no entity culling and the view boundary is not read from the dxf, it's determined by the bounding box of all the entities by matplotlib

**Reply by melihpeker-pointr** [reply]
Yes I get it now, it uses bbox for the size and margins. I thought that I could use the `export_mode` property of the RenderContext class for zoomed layouts as it states:
```
class RenderContext:    
    def __init__...
    Args:
    export_mode: Whether to render the document as it would look when exported (plotted) by a CAD application 
    to a file such as pdf, or whether to render the document as it would appear inside a CAD application. 
```

**Reply by mbway** [reply]
that only affects what layers are visible. There should probably be another parameter for whether to use layer boundaries / zoom but I haven't looked into that much.

**Reply by melihpeker-pointr** [reply]
Oh, ok I see. Thanks for the help.
